### PR TITLE
feat: use the new template load mechanism in ipyvue

### LINKED ISSUE #676: BUG: vue_destroy_viewer_item called twice on destroy event
https://github.com/spacetelescope/jdaviz/blob/e0efb4f2f7703b869437bd03feb9701b85a93a0b/jdaviz/app.py#L867-L870

To see it in action:

1. Add a `print(cid)` in that method.
2. Install `jdaviz` with the change above.
3. Start a viewer (say, Imviz).
4. Click "x" on the viewer.
5. See the `print` called twice for the same `cid`. 

I see threads like https://stackoverflow.com/questions/57211521/vue-component-being-destroyed-twice and https://stackoverflow.com/questions/41879836/vue-js-method-called-multiple-times-using-emit-and-on-when-it-should-only-be-c but none of the solutions make sense to me the way I see `jdaviz/container.vue` implemented.

Maybe @mariobuikhuizen can advise? Thanks!

### PR DESCRIPTION
**Pull request by mariobuikhuizen**
### Description

By using the new template loading mechanism in ipyvue, hot reloading (seeing changes to vue templates without a kernel restart) is now possible by adding the following code to the notebook:
```python
import ipyvue
ipyvue.watch('../jdaviz')
``` 

A simple demo:
https://user-images.githubusercontent.com/46192475/135246187-4c0f6163-cbcf-4e93-9ef0-5cbadfe81f41.mp4

Also fix #676 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/app.py [hunk 1]
```diff
@@ -105,6 +108,9 @@ class ApplicationState(State):
     tray_items = ListCallbackProperty(
         docstring="List of plugins displayed in the sidebar tray area.")
 
+    tray_open = CallbackProperty(
+        [], docstring="The plugin(s) opened in sidebar tray area.")
+
```
Comment (kecnry) [inline jdaviz/app.py:113]: Should we rename the `drawer` CallbackProperty (and reorder it with the others so it's not isolated at the top)?  On first read I couldn't tell the difference until I looked to see where/how they were both used, so perhaps `drawer` -> `tray_open` and `tray_open` -> `tray_items_open`?

File: jdaviz/app.py [hunk 2]
```diff
@@ -938,12 +937,16 @@ def remove(stack_items):
                 if len(stack.get('children', [])) > 0:
                     stack['children'] = remove(stack['children'])
 
+            empty_stacks = [s for s in stack_items
+                            if not s['viewers'] and not s.get('children')]
+            for empty_stack in empty_stacks:
+                stack_items.remove(empty_stack)
+
             return stack_items
```
Comment (kecnry) [inline jdaviz/app.py:945]: Can this all be simplified to a single list comprehension?  Or would that limit readability?

File: CHANGES.rst [hunk 3]
```diff
@@ -33,6 +33,7 @@ Specviz
 
 Bug Fixes
 ---------
+- vue_destroy_viewer_item called twice on destroy event #676
```
Comment (pllim) [inline CHANGES.rst:36]:
```suggestion

- ``vue_destroy_viewer_item`` called twice on destroy event. [#676, #913]
```

File: CHANGES.rst [hunk 4]
```diff
@@ -48,7 +49,7 @@ Specviz
 
 Other Changes and Additions
 ---------------------------
-
+- The new template load system in ipyvue is used, which enables hot reload.
```
Comment (pllim) [inline CHANGES.rst:52]:
```suggestion

- The new template load system in ``ipyvue`` is used, which enables hot reload. [#913]
```

File: jdaviz/app.py (hunk 1 above)
Comment (mariobuikhuizen) [reply jdaviz/app.py:113]: Yes, `tray_items_open` is a better name. The renaming of drawer should not go into this PR IMO.

Comment (pllim) [reply jdaviz/app.py:113]: Can we rename stuff without breaking stuff? Some of this `Application` setup is pretty fragile.

File: jdaviz/utils.py [hunk 5]
```diff
@@ -1,42 +1,7 @@
-import os
 import time
 import threading
 from collections import deque
```
Comment (pllim) [inline jdaviz/utils.py:4]: Don't delete `__all__`, just keep it empty.
```suggestion

__all__ = []
```

File: setup.cfg [hunk 6]
```diff
@@ -25,7 +25,7 @@ install_requires =
     glue-core>=1.2.2
     glue-jupyter>=0.10.1
     echo>=0.5.0
-    ipyvue>=1.4.1
+    ipyvue>=1.6
```
Comment (pllim) [inline setup.cfg:28]: Wasn't there another dependency you mentioned on Slack that is needed, or did I misunderstand?

**Review by pllim — COMMENTED** [review]
I don't grok the VueJS stuff, but I left a general review.

File: setup.cfg (hunk 6 above)
Comment (mariobuikhuizen) [reply setup.cfg:28]: Ah, yes, `watchdog` is an optional dependency. It's only needed when a dev wants to enable hot reloading.

Comment (pllim) [reply setup.cfg:28]: Should we list `watchdog` under "optional dependencies" or at least document this in the dev doc?

File: docs/installation.rst [hunk 7]
```diff
@@ -29,4 +29,13 @@ To install ``jdaviz`` for development or from source in editable mode::
...
    pip install -e .
 
+To enable the hot reloading of vue templates, install `watchdog`::
+
+   pip install watchdog
+
+And to the top of a notebook add::
+
+   import ipyvue
+   ipyvue.watch('../jdaviz')
```
Comment (kecnry) [inline docs/installation.rst:39]: The following works in general (including when using an installed `jdaviz` without having to worry about getting the path right manually):

```
import jdaviz
import os
import ipyvue
ipyvue.watch(os.path.dirname(jdaviz.__file__))
```

Should we consider a `jdaviz.enable_hot_reloading()` which wraps the above and could also handle checking for the `watchdog` optional dependency?

**Review by kecnry — COMMENTED** [review]
Confirmed that this works on my end and is going to hugely improve frontend dev workflow!

It's probably out of scope for this PR... but is there a way to get this working in voila too?  I tried just adding the required lines at the top of the template notebook (with a switch in the CLI), but that didn't seem to do the trick.

File: docs/installation.rst (hunk 7 above)
Comment (mariobuikhuizen) [reply docs/installation.rst:39]: yeah, good idea 👍. Will add it.

I will also look into the Voila problem, I think I know what's causing it.

**Review by kecnry — APPROVED** [review]

**Review by rosteen — APPROVED** [review]
Consider my mind blown, very cool. Appears to work from my testing, approved.

**Pull request merged**